### PR TITLE
Refactor package scanner

### PR DESCRIPTION
**Pull request by juancarlospaco**
# Refactor the package scanner

**Checks for:**

- if the Git Repos actually exist.
- if the `*.nimble` file actually exist.
- if the `Alias` relate to an existent `Name`.
- Blacklisted tags, like `nim`, `nimrod`, etc.
- Empty strings on fields.
- Unwanted whitespaces on tags.
- Too much tags, too few tags.
- Packages with duplicated URLs.
- Enforce required keys.
- Maximum number of tags.
- Maximum length of tags.
- More precise checking of the JSON.
- Other minor cases observed on the current JSON.

**Other features:**

- Updated for Nim `0.19.0`.
- Everything can be Configured or Disabled from a `const`.
- Self-documented, now `nim doc` works (to attract more contributions).
- Works online/offline, ( `when defined(ssl)` it will go Online and check all repos).
- Uses proper `logging` instead of `echo`.
- Overall is an improvement over the previous one.

**To be defined:**

So this "Linter" finds more errors on the current JSON, I see 2 options:

- Enable all checking, fix the JSON errors.
- Disable some checking, keep the JSON with some errors.

So basically you can pick and choose on the Strictness.

**Feel free to use the Suggestions feature of GitHub to suggest improvements.**
:cat:

### CODE REVIEW

**Comment by dom96** [comment]
Things I see immediately: camelCase and also test shouldn't fail.

But this is great, thank you!

**Comment by juancarlospaco** [comment]
Tests are expected to fail, because of the mentioned imperfections on the current JSON, as explained above.
:slightly_smiling_face:

**Comment by genotrance** [comment]
What are the next steps to get this merged?

Having a report of all issues for every package will be helpful - then each package owner can be asked to update their package. If no update is provided in a stipulated amount of time, we can then disable it.

**Comment by wizzardx** [comment]
> Feel free to use the Suggestions feature of GitHub to suggest improvements.

I searched for but couldn't find the Suggestions GitHub feature.

I'll add a suggestion or idea over here instead (for now). Some kind of ongoing CI against dev and stable (and maybe stable-1) for the packages.

This might make more sense after Nim 1.0 comes out.

Some kind of auto-maintained record of "last known to be good" combination of packages and compiler versions would also be a very nice artifact to generate - I'm thinking along the lines of how Haskell's Stackage service works, and their "nightly" snapshots).

There's probably some overlap between this idea, and what the nimble.directory service already does. eg, if we look at NimBorg on there:

https://nimble.directory/pkg/NimBorg

**Comment by juancarlospaco** [comment]
@dom96 Can you check again please :question: 
is all updated for `0.20.0` and green. :slightly_smiling_face:

File: package_scanner.nim [hunk 1]
```diff
@@ -1,135 +1,367 @@
...
+## - Alias relate to Names
+## - Profanity Filter by Levenshtein Distance.
+## - Try to download the nimble file from the repo.
+## - Try to Git clone the repo (deletes folder after).
+## - Works online/offline
+##
+## Use
+## ---
+##
+## - Off-line: ``nim c -r package_scanner2.nim`` (Faster, wont use Internet).
```
Comment (dom96) [inline package_scanner.nim]: I'd rather this be an explicit option, silently disabling internet checks could cause problems down the road.

File: package_scanner.nim [hunk 2]
```diff
@@ -1,135 +1,367 @@
...
+    "public domain",
+    "lgplv3 or gplv2",
+    "openssl and ssleay",
+    "openldap",
+    "apache 2.0 or gplv2",
+    "mit or apache 2.0",
+    "lgpl with static linking exception",
+  ]  ## All valid known licences for Nimble packages, on lowercase.
+when checkProfanity:
+  const profanitiesStr = readFile(currentSourcePath().parentDir / "profanities.txt").normalize.splitLines
```
Comment (dom96) [inline package_scanner.nim]: This is cool, but I don't think it's necessary, in fact if people want to call their package "fuck" I have no problem with that. I think humans can reject anything more offensive.

File: package_scanner.nim [hunk 3]
```diff
@@ -1,135 +1,367 @@
...
+addOutputFormatter(defaultConsoleFormatter())
+addOutputFormatter(report)
+when checkProfanity:
+  let profanities = filterIt(profanitiesStr, it.len > 0 and not(it.startsWith"#"))
+
+proc handler() {.noconv.} =
+  quit("CTRL+C Pressed, package_scanner is shutting down, Bye.")
+setControlCHook(handler)
+
+using url, name: string
```
Comment (dom96) [inline package_scanner.nim]: I personally dislike this construct, we shouldn't be using it.

File: package_scanner.nim [hunk 4]
```diff
@@ -1,135 +1,367 @@
...
+  rubyCmd = "ruby ./validate_json.rb"
+  defaultFilePermissions = {fpUserWrite, fpUserRead, fpGroupWrite, fpGroupRead, fpOthersRead}
+  packagesFilePath = currentSourcePath().parentDir / "packages.json"
+  packagesJsonStr = readFile(packagesFilePath)
+  hostsSkip = [
+    "https://bitbucket",
+    "https://gitlab.3dicc",
+    "https://mahlon@bitbucket",
+    "https://notabug",
+  ] ## Hostnames to skip checks, for reliability, wont like direct HTTP GET?.
```
Comment (dom96) [inline package_scanner.nim]: These are strange URLs, they don't contain a real hostname. I assume you are parsing them in some custom way? Why not make this more clear with:

```nim
hostsSkip = [
  "bitbucket.org",
  "gitlab.3dicc",
  ...
]
```

File: package_scanner.nim [hunk 5]
```diff
@@ -1,135 +1,367 @@
...
+    "apache license 2.0 or mit",
+    "fontconfig",
+    "gnu lesser general public license v2.1",
+    "public domain",
+    "lgplv3 or gplv2",
+    "openssl and ssleay",
+    "openldap",
+    "apache 2.0 or gplv2",
+    "mit or apache 2.0",
+    "lgpl with static linking exception",
```
Comment (dom96) [inline package_scanner.nim]: We need to be more strict and possibly implement a license parser which understands things like "MIT or Apache 2.0" in the long run.

File: package_scanner.nim [hunk 6]
```diff
@@ -1,135 +1,367 @@
...
+  ]  ## All valid known licences for Nimble packages, on lowercase.
```
Comment (dom96) [inline package_scanner.nim]: Nitpick: I prefer these to be written as appropriate, i.e. uppercase for acronyms and properly capitalised. You could programmatically pre-process this list and make it all lower-case.

File: package_scanner.nim [hunk 7]
```diff
@@ -1,135 +1,367 @@
...
+            nimble_nonexistent.add url
+
+      # Warn or Assert the possible errors at the end.
+      if nonexistent.len > 0 and allowBrokenUrl:
+        warn "Missing repos list:\n" & nonexistent.join("\n  ")
+        warn "Missing repos count: " & $nonexistent.len & " of " & $pckgsList.len
+      else:
+        doAssert nonexistent.len == 0, "Missing repos: Broken Packages."
+      if nimble_nonexistent.len > 0 and allowMissingNimble:
+        warn "Missing Nimble files:\n" & nimble_nonexistent.join("\n  ")
```
Comment (dom96) [inline package_scanner.nim]: Nitpick: please stick to camelCase throughout.

File: package_scanner.nim (hunk 1 above)
Comment (juancarlospaco) [reply package_scanner.nim]: Its not Silent, it prints a Hint using the Hint Pragma with explicit message.
Note that by default it runs with with SSL then it goes online.
But I get the idea, you want another additional `define` :grey_question:  :thinking:

File: package_scanner.nim (hunk 2 above)
Comment (juancarlospaco) [reply package_scanner.nim]: is compile time disabled, if some day want to enable it. :slightly_smiling_face:

File: package_scanner.nim (hunk 3 above)
Comment (juancarlospaco) [reply package_scanner.nim]: Why tho :grey_question: :thinking: Is just what the manual says, is something broken with the feature or something :grey_question: 
It wont say experimental nor deprecated.
Not like super important, I will remove it, just curious.

File: package_scanner.nim (hunk 5 above)
Comment (juancarlospaco) [reply package_scanner.nim]: I agree, but I am not secure how to implement that, maybe for next PR.

File: package_scanner.nim (hunk 7 above)
Comment (juancarlospaco) [reply package_scanner.nim]: u right

File: .gitignore [hunk 8]
```diff
@@ -0,0 +1,4 @@
+*.c
```
Comment (FedericoCeratto) [inline .gitignore]: Why those ignores?

File: .travis.yml [hunk 9]
```diff
@@ -1,26 +1,30 @@
...
       sh init.sh -y
-  - export PATH=$HOME/.nimble/bin:$PATH
 
 before_script:
   - set -e
   - set -x
 
 script:
-  - nim c -d:ssl -r package_scanner.nim && node ./validate_json.js
+  - nim c -d:ssl -d:release -r package_scanner.nim
```
Comment (FedericoCeratto) [inline .travis.yml]: -d:release slows down the compile time and does not improve significantly the tool run speed

File: package_scanner.nim [hunk 10]
```diff
@@ -1,135 +1,367 @@
...
 
-proc canFetchNimbleRepository(name: string, urlJson: JsonNode): bool =
-  # The fetch is a lie!
-  # TODO: Make this check the actual repo url and check if there is a
-  #       nimble file in it
-  result = true
-  var url: string
+func stripEndingSlash(url): string {.inline.} =
+  ## Strip the ending '/' if any else return the same string.
+  result = if url[^1] == '/': url[0 .. url.len - 2] else: url
```
Comment (FedericoCeratto) [inline package_scanner.nim]: url[0..^2]

File: package_scanner.nim [hunk 11]
```diff
@@ -1,135 +1,367 @@
...
 
-  if not urlJson.isNil:
-    url = urlJson.str
+func preprocessUrl(url, name): string =
+  ## Take **Normalized** URL & Name return Download link. GitHub & GitLab supported.
+  if likely(url.startswith("https://github.com/")):
```
Comment (FedericoCeratto) [inline package_scanner.nim]: The tool will spend most time doing networking and little else. Optimizing for branch prediction sounds quite unnecessary :)

File: package_scanner.nim [hunk 12]
```diff
@@ -1,135 +1,367 @@
...
 
-proc verifyAlias(pdata: JsonNode, result: var int) =
-  if not pdata.hasKey("name"):
-    echo "E: missing alias' package name"
-    result.inc()
+proc existsGitRepo(url): string =
+  ## Take **Normalized** URL try to Fetch the Git repo index page. Needs SSL.
+  if url.startswith("http"):
+    try:
+      doAssert client.get(url).status == $Http200 # Check that Repo Exists.
```
Comment (FedericoCeratto) [inline package_scanner.nim]: Nitpick: in theory setting assertions and catching them is not supposed to be used for control flow and you can easily do if .. raise

File: package_scanner.nim [hunk 13]
```diff
@@ -1,135 +1,367 @@
...
+    result = url  # SSH or other non-HTTP kinda URLs?
 
-  # TODO: Verify that 'alias' points to a known package.
 
-proc check(): int =
-  var name: string
-  echo ""
-  let pkg_list = parseJson(readFile(getCurrentDir() / "packages.json"))
-  var names = initSet[string]()
+suite "Packages consistency testing":
```
Comment (FedericoCeratto) [inline package_scanner.nim]: Unittest is not meant to test things outside code. It looks like it's testing the tool itself...

File: package_scanner.nim [hunk 14]
```diff
@@ -1,135 +1,367 @@
...
-      elif pdata["url"].str.normalize.startsWith("git://github.com/"):
-        echo "E: ", name, " has an insecure git:// URL instead of https://"
-        result.inc()
+  test "Check file permissions":
+    check getFilePermissions(packagesFilePath) == defaultFilePermissions
+
+  test "Check validate whole JSON by NodeJS":
+    check execCmd(nodeCmd) == 0
+
+  test "Check validate whole JSON by Python":
```
Comment (FedericoCeratto) [inline package_scanner.nim]: Testing if a JSON file is correct in 3 languages sounds a bit unneded

File: package_scanner.nim (hunk 2 above)
Comment (FedericoCeratto) [reply package_scanner.nim]: It's more code to be maintained...

File: package_scanner.nim (hunk 13 above)
Comment (juancarlospaco) [reply package_scanner.nim]: What would be the proper wording :grey_question:

File: package_scanner.nim (hunk 6 above)
Comment (juancarlospaco) [reply package_scanner.nim]: On the JSON theres weird casing, I always think that doing compare is better on lowercase,
this code is not mean to be seen by end user whatsoever.

Or you mean enforcing proper case on the JSON :grey_question: 
:thinking:

**Comment by FedericoCeratto** [comment]
We are starting to get into the infamous problem of library fragmentation that affects some other languages. I think we should encourage people not to publish tiny libraries. One way is to add a check on codebase size here or in Nimble and politely warn the developer where needed.

**Comment by juancarlospaco** [comment]
I understand the idea about a capitalization linter thing, and the idea about size checking,
but I think those can come on a future PR, the diff is already huge, is working and green,
so maybe merge this and more improvements can come on the future, on other PR.

:slightly_smiling_face:

File: package_scanner.nim (hunk 1 above)
Comment (dom96) [reply package_scanner.nim]: I would prefer a run-time argument. Also I want the default behaviour to be "online".

File: package_scanner.nim (hunk 6 above)
Comment (dom96) [reply package_scanner.nim]: > this code is not mean to be seen by end user whatsoever.

We are the end users. We use this scanner to tell us whether the newly added package is valid, this is just a style choice though so it's not a big deal. I don't see what JSON has to do with this.

File: package_scanner.nim (hunk 1 above)
Comment (juancarlospaco) [reply package_scanner.nim]: The default behavior to be "online".
No extra option required anywhere.
All other is for when a dev want to run it manually for whatever reason.

:slightly_smiling_face:

**Comment by juancarlospaco** [comment]
@dom96 Let me know any other feedback.

File: package_scanner.nim (hunk 1 above)
Comment (dom96) [reply package_scanner.nim]: No, the default is ``offline``. Default means "compiling with no options".

**Comment by dom96** [comment]
We still have unresolved feedback here. Please comment on each why you disagree with me or fix it.

**Comment by juancarlospaco** [comment]
@dom96 Check again, let me know if you see any problem.

**Comment by juancarlospaco** [comment]
@dom96 Let me know if something is still missing...
:thinking:

**Comment by xflywind** [comment]
I closed it because https://github.com/nim-lang/packages/pull/2027 is merged.
